File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Codec:
    
    # Encodes a tree to a single string.
    def serialize(self, root: Optional[TreeNode]) -> str:
        if not root:
            return 'N,'
        s = ""
        s += str(root.val) 
        s += ','

        s += self.serialize(root.left)
        s += self.serialize(root.right)

        return s
        
    # Decodes your encoded data to tree.
    def deserialize(self, data: str) -> Optional[TreeNode]: 
        self.i = 0;
        def dfs(data):
            if data[self.i] == 'N':
                self.i+=2
                return None
            s = ""
            while self.i < len(data) and data[self.i] != ',':
                s += data[self.i]
                self.i += 1
            self.i += 1;
            if not s:
                return None
            root = TreeNode(int(s)) 
            root.left = dfs(data)
            root.right = dfs(data)

            return root
        return dfs(data)

        

File: Data Structures & Algorithms/balanced-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        bool isBalanced = 1;
        dfs(root, isBalanced);
        return isBalanced;


    }

    int dfs(TreeNode* root, bool &isBalanced){
        if( root == nullptr){
            return 0;
        }

        if(abs(dfs(root->left, isBalanced) - dfs(root->right, isBalanced)) > 1){
            isBalanced = 0;
        }

        return 1 + max(dfs(root->left, isBalanced), dfs(root->right, isBalanced));
    }
    
};
